stubparser/src/org/checkerframework/stubparser/ast/visitor/DumpVisitor.java

```diff
@@ -221,9 +221,22 @@ private void printMembers(List<BodyDeclaration> members, Object arg) {
 
     private void printMemberAnnotations(List<AnnotationExpr> annotations, Object arg) {
         if (annotations != null) {
+            for (Iterator<AnnotationExpr> i = annotations.iterator(); i.hasNext();) {
+                AnnotationExpr a = i.next();
+                a.accept(this, arg);
+                if (i.hasNext()) {
+                    printer.print(" ");
+                }
+            }
+            printer.printLn();
+        }
+    }
+
+    private void printReceiverAnnotations(List<AnnotationExpr> annotations, Object arg) {

```

Comment:
Actually, there were already two loops, and there is a difference: either
the iterator or some ad-hoc method is necessary to make the first (or last)
item "special", in that there is no leading or trailing space.  It is true
that some of the commonality could be factored out, but I'm not convinced
it would make the code any clearer in this case.

-- Dan.


On Wed, Aug 12, 2015 at 3:58 PM, Werner Dietl <notifications@github.com>
wrote:

> In
> stubparser/src/org/checkerframework/stubparser/ast/visitor/DumpVisitor.java
> <https://github.com/typetools/checker-framework/pull/462#discussion_r36924849>
> :
>
> > @@ -221,9 +221,22 @@ private void printMembers(List<BodyDeclaration> members, Object arg) {
> >
> >      private void printMemberAnnotations(List<AnnotationExpr> annotations, Object arg) {
> >          if (annotations != null) {
> > +            for (Iterator<AnnotationExpr> i = annotations.iterator(); i.hasNext();) {
> > +                AnnotationExpr a = i.next();
> > +                a.accept(this, arg);
> > +                if (i.hasNext()) {
> > +                    printer.print(" ");
> > +                }
> > +            }
> > +            printer.printLn();
> > +        }
> > +    }
> > +
> > +    private void printReceiverAnnotations(List<AnnotationExpr> annotations, Object arg) {
>
> But why do you add so much duplication? You now have three for loops
> instead of one. If the only difference is whether there is a space before
> or not, this is possible a lot simpler.
> Also, some use extended for loops others use an iterator. This confuses
> readers and makes one look for some underlying difference - which isn't
> there.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/typetools/checker-framework/pull/462/files#r36924849>.
>